cmd/kaspaminer/mineloop.go

```diff
@@ -132,48 +125,65 @@ func handleFoundBlock(client *minerClient, block *externalapi.DomainBlock) error
 	return nil
 }
 
-func solveBlock(block *externalapi.DomainBlock, stopChan chan struct{}, foundBlock chan *externalapi.DomainBlock) {
-	targetDifficulty := difficulty.CompactToBig(block.Header.Bits())
-	headerForMining := block.Header.ToMutable()
+func mineNextBlock(mineWhenNotSynced bool) *externalapi.DomainBlock {
 	initialNonce := rand.Uint64() // Use the global concurrent-safe random source.
 	for i := initialNonce; i != initialNonce-1; i++ {
-		select {
-		case <-stopChan:
-			return
-		default:
-			headerForMining.SetNonce(i)
-			atomic.AddUint64(&hashesTried, 1)
-			if pow.CheckProofOfWorkWithTarget(headerForMining, targetDifficulty) {
-				block.Header = headerForMining.ToImmutable()
-				foundBlock <- block
-				return
-			}
+		block := getBlockForMining(mineWhenNotSynced)
+		targetDifficulty := difficulty.CompactToBig(block.Header.Bits())
+		headerForMining := block.Header.ToMutable()
```

Comment:
I think the above three lines should be outside of the loop.

Reply:
No, because you can get different block for each nonce. This reduces the hash rate by ~20%, but it's just a CPU example miner, so I think it's not a big deal.

Reply:
In that case, please:
Rename initialNonce to nonce
Unfancify the `for` declaration (it's just an infinite loop)
Increment the nonce in the loop's body 
Add a comment explaining what's going on here